frontends/common/resolveReferences/resolveReferences.h

```diff
@@ -29,6 +29,12 @@ enum class ResolutionType { Any, Type, TypeVariable };
 
 /// Visitor mixin for looking up names in enclosing scopes from the Visitor::Context
 class ResolutionContext : virtual public Visitor, public DeclarationLookup {
+    mutable std::unordered_map<const IR::INamespace *, std::vector<const IR::IDeclaration *> *>
```

Comment:
What requires the addition of mutable? This doesn't seem right to me.

Reply:
As @vlstill mentioned, `mutable` members are often used to avoid making semantically `const` methods non-const. But @asl (the original author) may add additional comment here.

Reply:
Well, the methods are `const` for a reason. Adding mutable defeats the purpose. I understand using mutable for threaded implementations or counters, but here it appears like flawed design to me.

Reply:
This is a pretty standard approach to use `mutable` for the purposes of memoization. Essentially we are capturing the return value of `const` method and next time it is called with the same arguments we used the captured result.

So, capturing result is a stateful method and therefore we'd either need to remove const (that is not semantically correct) or make the cache `mutable`. The common design patter to use the latter approach.

Reply:
Yep, this is pretty standard. The function is still semantically const (calling just const functions does not change their results for the same inputs), the only thing that can be observed is timing. If we disregard threads that is :-). But luckily there are no threads except for GC, which is no-issue here (if someone were to compute something in parallel they would not be able to share `ReferenceMap`).

I believe the standard library internally uses a stronger contract for `const` that mandates that `const` functions are safe to be called in parallel, but I am not 100 % sure about that.

Reply:
> I believe the standard library internally uses a stronger contract for const that mandates that const functions are safe to be called in parallel, but I am not 100 % sure about that.

Yes, we do have this at least in libc++ (have no idea about libstdc++, did not touch it for ages). Introduction of threads would require lots of changes here and there as there lots of data structures that are shared inside p4c. It would require careful review and mutexes here and there. Should this be necessary we can introduce some generic "memoize" (similar to what's in Folly for example) wrapper, but I think it should be done separately.

Reply:
Yep, I've mentioned threads just as a side note :-). That is a whole new level of complexity that I hope we can get by without.

Reply:
Personally, I think being forced to use `mutable` is not a good sign. There is typically an alternative, phased solution that involves pre-computation available. While usage of `mutable` is standard there are many things standard in C++, which is a big mess of a language. 

For P4C itself, maintainability and correctness used to take precedence over performance. With that the compiler's  C++philosophy has been to implement a less elegant/performant, but easily understandable solution. In cases, where that was not the case we are often paying the prize. For instance there are several complex passes in the compiler such as the Predication or LoopsUnrolling pass, which are broken in parts but no one has the time or energy to fix them. Ultimately, all of this becomes maintenance burden.

If we go with this option it would be great to add some more documentation to ReferenceMap/ResolveReferences and highlight that the appropriate functions are using caching which depends on `mutable`.

Reply:
> For P4C itself, maintainability and correctness used to take precedence over performance. With that the compiler's C++philosophy has been to implement a less elegant/performant, but easily understandable solution. In cases, where that was not the case we are often paying the prize. 

These are all important trade-offs that should be made. Wearing my LLVM hat on, I am all for maintainability and clearness of implementation. Note, however, that compiler itself is a tool that is used by many people all the day. And as far as I can see, there are some areas where p4c performance could definitely be better. I am seeing lots of cases where not very complex P4 input causes massive compiler slowdown. And here we are talking not about seconds of compilation time, but minutes and tens of minutes.

For example, in the original downstream testcase this change dropped the compilation time from 129 seconds down to 34. These days we are trying to tackle cases when p4c itself takes 10-15 minutes to run effectively causing very poor developer experience.

There are lots of factors that contribute to this poor performance. Some due to easy to fix corner cases, but others are due to more widespread overall design inefficiencies. And if we are talking about 3-4x compiler speedup as compared to some additional implementation complexity, then for me the tradeoff is clear. Hope the p4c contributors share my view on the things here.

Certainly, one could try to rewrite the entire implementation to get rid of quadratic behavior in this particular case. Maybe we will have time for this one day. For now we are upstreaming the downstream patches that we know to work reasonable well. We are certainly trying to cleanup them as much as possible.

Reply:
I agree, the performance of the compiler definitely needs work. It has been a long-standing problem and everyone benefits from the optimizations, so they are definitely welcome! 

My concern was mostly regarding maintenance on the open-source side, as there are few to none people claiming ownership on the involved code these days. If things go wrong the person responsible may not be reachable. And things such as `mutable` can make debugging very tricky as const-qualified functions may not behave as one would expect. 

Thankfully, we do have fairly expansive testing available. However, issues can still slip by. It is good to know that this code was used extensively downstream and these are patches. I was missing this context, for example.

Reply:
None of the tests in the repository tests the multithreaded compiler, which is used extensively in the production environment for backtracking. Ideally @ChrisDodd would weigh on this.

Reply:
@mihaibudiu This looks like a problem. Can the tests be added somehow?

Reply:
I personally don't have any multithreaded tests to add. Maybe @ChrisDodd has.

Reply:
Are we sure that this will cause a problem with the multithreaded compiler, or are we just waiting for @ChrisDodd to comment?

Reply:
We don't know

Reply:
This might happen if two different threads would share a common `ResolutionContext`. Given that the latter is `Visitor` itself, I would consider it highly unlikely...

Reply:
The idea behind supporting multithreading in the compiler is mostly by having the IR always be read-only, so multiple threads can traverse and visit it simultaneously.  Visitors that don't need to store anything can also be read-only, but any visitor that keeps around traversal state will require a separate instance of the visitor for each thread.  So similarly limiting a ResolutionContext to a single thread is also not too bad.